File: examples/julia/inspect_tick_events.jl
using HDF5, Dates, Statistics, StatsBase          # Load core libraries for HDF5 I/O, date/time, stats, and statistical sampling
using GRUtils                                     # Plotting backend

path = joinpath(homedir(), "iex.h5")              # Construct path to HDF5 container in user's home directory
fd = h5open(path, "r")                            # Open HDF5 file for reading
dates = read(fd["/trading_days.txt"])             # Read list of trading days (as strings or dates)
instruments = read(fd["/instruments.txt"])        # Read list of instrument names

stats = fd["/stats"]                              # Navigate into the stats group
day = last(dates)                                 # Select the last available trading day

N = Int.(read(stats[day]["trade_count"]))         # Read trade counts for that day and convert to Int
sort(N, rev=true)                                 # Sort trade counts descending (not assigned here)

cutoff_limit = 10                                 # Show top 10 most active instruments by trade count
top_indices = sortperm(N, rev=true)               # Get permutation of indices sorted by descending trade count
N[top_indices]                                    # Display sorted trade counts

instruments[top_indices][1:10]                    # Show top 10 most traded instrument names
ticks = read(fd["irts/" * day])                   # Read all tick events for the selected day

i = top_indices[3] - 1                            # Select the 3rd most traded instrument (adjusting for 0-based contract_id)

# Count tick types for this instrument (by flags):
nbids   = count(t -> t.contract_id == i && t.flags == 0x0001, ticks)  # Number of BID events
ntrades = count(t -> t.contract_id == i && t.flags == 0x0002, ticks)  # Number of TRADE events
nasks   = count(t -> t.contract_id == i && t.flags == 0x0004, ticks)  # Number of ASK events
nall    = count(t -> t.contract_id == i, ticks)                       # Total number of events for this instrument
nbids + ntrades + nasks == nall                                      # Sanity check: part sums should equal total

A = filter(t -> t.contract_id == i && t.flags == 0x0002, ticks)       # Extract only TRADE events for the instrument

times_ns = [t.time for t in A]                                        # Get trade timestamps in nanoseconds (UInt64 UNIX time)
times_dt = unix2datetime.(times_ns ./ 1e9)                            # Convert nanosecond UNIX timestamps to DateTime
prices = [t.price for t in A]                                         # Extract trade prices

plot(prices)                                                          # Plot the price time series
instruments[i]                                                        # Return instrument name (Julia indexing)

contract_ids = [t.contract_id for t in ticks]                         # Collect all contract IDs for further analysis
close(fd)                                                             # Close the HDF5 file

File: examples/python/inspect_tick_events.py
import os                                              # Filesystem utilities
import h5py                                            # HDF5 reader
import numpy as np                                     # Numerical operations
import matplotlib.pyplot as plt                        # Plotting
from datetime import datetime                          # Time conversion

def unix2datetime_ns(ns_array):                        # Convert UNIX ns timestamps to datetime
    return [datetime.utcfromtimestamp(ns * 1e-9) for ns in ns_array]

path = os.path.expanduser("~/iex.h5")                  # Path to HDF5 file
fd = h5py.File(path, "r")                              # Open file for reading

dates = [d.decode() for d in fd["/trading_days.txt"][:]]       # Read trading day list
instruments = [i.decode() for i in fd["/instruments.txt"][:]]  # Read instrument names

stats = fd["/stats"]                                   # Navigate into stats group
day = dates[-1]                                        # Select last trading day

N = stats[day]["trade_count"][:].astype(int)           # Read trade counts and convert to int
top_indices = np.argsort(N)[::-1]                      # Indices sorted by descending count

cutoff_limit = 10                                      # Number of top instruments to show
top_names = [instruments[i] for i in top_indices[:cutoff_limit]]  # Top instrument names
top_counts = N[top_indices[:cutoff_limit]]             # Top instrument trade counts

ticks = fd[f"/irts/{day}"][:]                          # Load tick events for selected day

i = top_indices[4]                                     # Select 6th most traded instrument

nbids = np.count_nonzero((ticks["contract_id"] == i) & (ticks["flags"] == 0x0001))   # Count BID events
ntrades = np.count_nonzero((ticks["contract_id"] == i) & (ticks["flags"] == 0x0002)) # Count TRADE events
nasks = np.count_nonzero((ticks["contract_id"] == i) & (ticks["flags"] == 0x0004))   # Count ASK events
nall = np.count_nonzero(ticks["contract_id"] == i)                                   # Total events

assert nbids + ntrades + nasks == nall                        # Sanity check

trades = ticks[(ticks["contract_id"] == i) & (ticks["flags"] == 0x0002)]  # Extract trade events

times_ns = trades["time"]                                    # Extract timestamps
times_dt = unix2datetime_ns(times_ns)                        # Convert to datetime
prices = trades["price"]                                     # Extract prices

plt.figure(figsize=(10, 4))                                  # Create figure
plt.plot(times_dt, prices)                                   # Plot prices vs. time
plt.title(f"Trade prices for {instruments[i]}")              # Plot title
plt.xlabel("Time")                                           # X axis label
plt.ylabel("Price")                                          # Y axis label
plt.grid(True)                                               # Add grid
plt.tight_layout()                                           # Improve layout
plt.savefig("plot.png")                                      # Save plot to file

contract_ids = ticks["contract_id"].astype(int)              # Extract all contract IDs
